CLN: Removed the flavor parameter in DataFrame.to_sql

gfyoung opened the pull request:
Deprecated in `0.14.0`, so way, way overdue.

gfyoung committed:
CLN: Removed the flavor parameter in DataFrame.to_sql

[ci skip]

jorisvandenbossche commented:
@gfyoung We cannot just remove the flavor keyword. It was only deprecated for 'mysql', not for 'sqlite' (wihch is a pitty ..)

gfyoung commented:
But the argument accepts only one value (`sqlite`), so why not remove it?

jorisvandenbossche commented:
No, we are certainly not planning that. And in retrospect, it's a pitty that we didn't deprecate the full flavor keyword. But we have to deal with it now. I think we should give a nice error/warning message in case of `flavor='sqlite'`. So either deprecated that, or 'remove' it but still capture it's usage to display a more informative error message.

gfyoung commented:
@jorisvandenbossche : How does adding a `**kwargs` parameter sound?  Then we can still accept it and warn about it if the user passes it in.

jorisvandenbossche commented:
@gfyoung That sounds good!

gfyoung commented:
@jorisvandenbossche : added `*args` and `**kwargs` along with a validation function AND skipped all of the test suite that relied on the `flavor="mysql"` connection.  However, I would like to add a test to make sure the `flavor` parameter is properly deprecated without these functions getting abused, but I just want to make sure these initial changes appeal the Travis.

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -1130,8 +1130,8 @@ def to_msgpack(self, path_or_buf=None, encoding='utf-8', **kwargs):
-    def to_sql(self, name, con, flavor='sqlite', schema=None, if_exists='fail',
-               index=True, index_label=None, chunksize=None, dtype=None):
+    def to_sql(self, name, con, schema=None, if_exists='fail', index=True,
+               index_label=None, chunksize=None, dtype=None, *args, **kwargs):
```
I don't think the `*args` are needed here. If people were using positional kwargs, the current deprecation will not work anyways (as `schema` will catch the flavor)

jorisvandenbossche replied:
Alternative is to keep `flavor` in the signature, but change it to None, so we can detect when people set it to `'sqlite'` themselves. That solves the positional kwarg problem, but of course pollutes the signature a bit.

gfyoung replied:
Argh...good point.  I think I'll just remove the '*args' parameter and not pollute the signature.

gfyoung commented:
@jorisvandenbossche : Updated the signatures and Travis is still giving the green light.  Ready to merge if there are no other concerns.

gfyoung commented:
@jorisvandenbossche : Updated the signatures, and Travis is still giving the green light.  Ready to merge if there are no other concerns.

gfyoung commented:
@jreback , @jorisvandenbossche : Updated the signatures, and Travis is still giving the green light.  Ready to merge if there are no other concerns.

gfyoung commented:
@jreback , @jorisvandenbossche : Updated the signatures, and Travis is still giving the green light.  Ready to merge if there are no other concerns.

jorisvandenbossche commented:
@gfyoung yeah, I saw your ping, but I also have other work :-)

I worry a bit about the usage of `flavor` as positional arg .. We actually used it like at some places in the tests I saw in the diff. But it is difficult to judge whether this is used like that by users (a quick stackoverflow search didn't turn anything up). 
It would be the safer option to leave it in the signature I think until we actually remove it.

gfyoung commented:
Yeah...that is a fair point.  Original purpose was to just remove the 'mysql' option, so let me scale this PR back down to that.

jorisvandenbossche commented:
@gfyoung Yes, but I think it would still be good to deprecated the flavor arg, so we can remove it later on. So if you want to do that, I don't it is a large change from what you have now, it's just adding it back in the signature and replacing the kwarg validator with only the deprecation part of it

gfyoung commented:
@jorisvandenbossche : Updated the signature and put the `flavor` parameter back.  However, it was tricky to issue warnings because `flavor` is a keyword with default value as `"sqlite"` in MOST cases.  As we don't want to punish users who don't pass it in, and since it has no effect anymore, I changed the default value in ALL signatures to `None`.  Travis is happy with this change, so ready to merge if there are no other concerns.

gfyoung commented:
@jreback, @jorisvandenbossche : rehashed and Travis is still happy - ready to merge if there are no other concerns.

jorisvandenbossche commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -492,6 +492,7 @@ Deprecations
 - ``Categorical.reshape`` has been deprecated and will be removed in a subsequent release (:issue:`12882`)
 - ``Series.reshape`` has been deprecated and will be removed in a subsequent release (:issue:`12882`)
 
+- ``DataFrame.to_sql()`` has deprecated the ``flavor`` parameter (:issue:`13611`)
```
Can you emphasize that is is a superfluous keyword (so no reason to use it), since sqlite is the only supported connection when not using sqlalchemy?

gfyoung replied:
Done.

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -1157,10 +1157,8 @@ def to_sql(self, name, con, flavor='sqlite', schema=None, if_exists='fail',
-            'mysql' is deprecated and will be removed in future versions, but
-            it will be further supported through SQLAlchemy engines.
+        flavor : 'sqlite', default None
+            DEPRECATED: this parameter will be removed in a future version
```
similar comment here, maybe add "because only 'sqlite' is supported"

gfyoung replied:
Done.

jorisvandenbossche commented on pandas/io/sql.py:
```diff
@@ -41,6 +41,22 @@ class DatabaseError(IOError):
+    if flavor is not None:
+        if flavor == 'sqlite':
+            warnings.warn("the 'flavor' parameter is deprecated "
+                          "and will be removed in a future version.",
```
and the same here. Something like: "it is not needed to specify this, because 'sqlite' is the default and only supported flavor when not using sqlalchemy" (so users don't think using sqlite itself is deprecated)

gfyoung replied:
Done.

jorisvandenbossche commented on pandas/io/tests/test_sql.py:
```diff
@@ -2425,44 +2308,68 @@ def clean_up(test_table_to_drop):
+    """
+
+    con = 1234  # don't need real connection for this
+    funcs = ['SQLiteDatabase', 'pandasSQL_builder']
```
I would rather test the top-level `DataFrame.to_sql` (and then check_stacklevel can be set to True)

jorisvandenbossche replied:
although SQLiteDataBase can also be tested, but `pandasSQL_builder` is not needed I think

gfyoung replied:
to_sql isn't just the only function that calls pandasSQL_builder though that I changed - seems more reasonable to test the function underlying all the ones whose signatures I changed IMO.

Also, why is pandasSQL_builder not need?  It has no underscore prefix --> not an internal function technically.

jorisvandenbossche replied:
> Also, why is pandasSQL_builder not needed? It has no underscore prefix --> not an internal function technical

But is should actually have one :-)

But it' true that it is used by more than to_sql, so OK leave it like that!

gfyoung replied:
@jorisvandenbossche : Ah, then I'll put up a PR to underscore it as a follow-up.

jorisvandenbossche commented:
@gfyoung Looks good! (using flavor=None is perfect for detecting if user has specified it themselves) Few minor comments.

gfyoung commented:
@jorisvandenbossche : made the requested doc changes, so ready to merge if there are no other concerns.

gfyoung commented:
@jorisvandenbossche : any updates on this?

jorisvandenbossche commented:
Yes, merging! Thanks a lot
